ts-rs/src/typelist.rs

```diff
@@ -0,0 +1,47 @@
+use std::any::TypeId;
+use std::marker::PhantomData;
+
+use crate::TS;
+
+pub trait TypeVisitor: Sized {
+    fn visit<T: TS + 'static + ?Sized>(&mut self);
+}
+
+
+pub trait TypeList: Copy + Clone {
+    fn push<T: TS + 'static + ?Sized>(self) -> impl TypeList {
+        (self, (PhantomData::<T>,))
+    }
+    fn extend(self, l: impl TypeList) -> impl TypeList {
+        (self, l)
+    }
+
+    fn contains<C: Sized + 'static>(self) -> bool;
+    fn for_each(self, v: &mut impl TypeVisitor);
+}
+
+impl TypeList for () {
+    fn contains<C: Sized>(self) -> bool { false }
+    fn for_each(self, _: &mut impl TypeVisitor) {}
+}
```

Comment:
An empty `TypeList` is just `()`. That's why `TS::dependency_types() -> impl TypeList` can have a default implementation with an empty body, since that evaluates to `()`

Reply:
Nice!